rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,188 @@
+So we will report the violated assertions as `UNDETERMINED` instead of `FAILED`.
+
+A question about how do we print the synthesized loop contracts when users request is discussed in **Open question**.
+
+## Detailed Design
+The synthesizer ```goto-synthesizer``` is implemented in the repository of `CBMC`, takes as input a goto binary, and outputs a new goto binary with the synthesized loop contracts applied.
+Currently, Kani invokes `goto-instrument` to instrument the goto binary `main.goto` into a new goto binary `main_instrumented.goto`, and then invokes ```cbmc``` on `main_instrumented.goto` to get the verification result.
+The synthesis will happen between calling `goto-instrument` and calling ```cbmc```.
+That is, we invoke ```goto-synthesizer``` on ```main_instrumented.goto``` to produce a new goto binary ```main_synthesized.goto```, and then call ```cbmc``` on `main_synthesized.goto` instead. 
+
+When invoking ```goto-synthesizer```, we pass the following parameters to it with the flags built in `goto-synthesizer`:
+- the resource limit of the synthesis;
+- the solver options to specify what SAT solver we use to verify invariant candidates.
+
+The enumerator used in the synthesizer enumerates candidates from the language of the following grammar template
+```
+NT_Bool -> NT_Bool && NT_Bool | NT_int == NT_int 
+            | NT_int <= NT_int | NT_int < NT_int 
+            | SAME_OBJECT(terminals_ptr, terminals_ptr)
+            
+NT_int  -> NT_int + NT_int | terminals_int | LOOP_ENTRY(terminals_int)
+            | POINTER_OFFSET(terminals_ptr) | OBJECT_SIZE(terminals_ptr)
+            | POINTER_OFFSET(LOOP_ENTRY(terminals_ptr)) | 1
+```
+where `terminals_ptr` are all pointer variables in the scope, and `terminal_int` are all integers variables in the scope.
+
+We use the following example to illustrate how the synthesizer works.
+```rust
+#[kani::proof]
+fn main() {
+    let mut y: i32 = kani::any_where(|i| *i > 0);
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+As there is only one variable `y` in the scope, the grammar template above will be instantiated to the following grammar
+```
+NT_Bool -> NT_Bool && NT_Bool | NT_int == NT_int 
+            | NT_int <= NT_int | NT_int < NT_int 
+NT_int  -> NT_int + NT_int | y | LOOP_ENTRY(y) | 1
+```
+The synthesizer will enumerate candidates derived from `NT_Bool` in the following order.
+```
+y == y
+y == LOOP_ENTRY(y)
+y == 1
+...
+1 <= y + 1
+...
+```
+The synthesizer then verifies with CBMC if the candidate is an inductive invariant that can be used to prove the post-condition `y == 0`.
+For example, the candidate `y == y` is verified to be an inductive invariant, but cannot be used to prove the post-condition `y == 0`.
+The candidate `y == 1` is not inductive.
+The synthesizer rejects all candidates until it finds the candidate `1 <= y + 1`, which can be simplified to `y >= 0`.
```

Comment:
Does the synthesizer call `cbmc` every time it finds an inductive invariant?

Reply:
Yes, the synthesizer run `CBMC` every time it needs to check if a candidate is inductive.

Reply:
And if CBMC says it is inductive, will it call CBMC again to check the program?

Reply:
They are checked in the same call to CBMC.
For a candidate invariant, we apply it to the program and run CBMC to verify the programs including the inductiveness checks and all other checks.
If all checks pass, the candidate invariant is good to return.
If inductive checks pass, the candidate invariant is inductive, we can keep it as the strongest inductive invariant we have seen.
If inductive checks fail, we discard the candidate.

Reply:
I see, thanks for explaining. Can you clarify this a bit in the RFC as it wasn't clear to me?

Reply:
Sure.